### ISSUE TITLE
Uncaught TypeError: Cannot read property 'id' of undefined

### ISSUE THREAD

**Issue by relupselav** [atom/atom#15441]
Steps to reproduce by @mangecoeur https://github.com/atom/atom/issues/15441#issuecomment-325651325

[test.zip](https://github.com/atom/atom/files/1260242/test.zip)

Steps:

1. unzip the file and open test.md in atom
1. Make the Atom Window narrow
1. trigger `preview-inline` on BOTH image links
1. Scroll up and down
-> Error seems to be triggered when the top of the block decoration of the first decoration enters or exits the viewable window.
1. Close the first preview
-> Error seems to be triggered when the top of the block decoration of the remainding decoration enters or exits the viewable window. NOTE that if you close the second preview, but keep the first, the error is not triggered `¯\_(ツ)_/¯`
1. Close both preview
-> The error no longer appears.

**Atom**: 1.19.3 x64
**Electron**: 1.6.9
**OS**: Mac OS X 10.12.6
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: Cannot read property 'id' of undefined

```
At /Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:1185

TypeError: Cannot read property 'id' of undefined
    at TextEditorComponent.addBlockDecorationToRender (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:1185:67)
    at TextEditorComponent.addDecorationToRender (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:1026:22)
    at decorationsByMarker.forEach (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:995:20)
    at Map.forEach (native)
    at TextEditorComponent.queryDecorationsToRender (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:990:31)
    at TextEditorComponent.updateSyncBeforeMeasuringContent (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:381:16)
    at TextEditorComponent.updateSync (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:276:16)
    at didDrag (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:1806:20)
    at get_window.requestAnimationFrame (/Users/valentinaspuler/Downloads/Atom.app/Contents/Resources/app/src/text-editor-component.js:1881:17)
```

### Commands

```
     -1:16 application:add-project-folder (input.hidden-input)
  7x -0:55.5.0 core:backspace (input.hidden-input)
```

### Non-Core Packages

```
aligner 1.2.1 
aligner-javascript 1.1.0 
atom-beautify 0.29.24 
auto-indent 0.5.0 
busy-signal 1.4.1 
color-picker 2.2.5 
emmet 2.4.3 
highlight-selected 0.13.1 
intentions 1.1.2 
linter 2.1.4 
linter-ui-default 1.2.4 
pigments 0.39.1 
```

**Comment by Ben3eeE** [comment]
Thanks for the report!  Can you confirm a few things for us?

* what were you doing at the time of the error?
* is the error is reproducible?
* if reproducible, do you see the error in safe mode (atom --safe)?

**Comment by mangecoeur** [comment]
I can also confirm this issue. It is triggered for example when using the preview-inline plugin, which creates block-decoration elements containing images. It is triggered when scrolling with the mouse cursor held over the block-decoration elements.

This issue was raised and fixed previously, I suspect that the editor/buffer rewrite has re-introduced it. I believe that it was caused by the editor making assumptions about every element under the mouse having certain properties, which would not be true in some cases including image and svg elements.

**Comment by Ben3eeE** [comment]
@mangecoeur I am unable to reproduce on Atom 1.19.4 here is what I did.

1. Installed the `preview-inline` plugin from the `settings-view`
1. Open Atom in the `atom/atom` repository
1. Open `README.md`
1. On the first line which contains the Atom logo trigger `preview-inline` with <kbd>Ctrl-Alt-p</kbd>
1. With the mouse cursor held over the preview image attempt to scroll up and down

Expected and Actual: The editor scrolls.

What is it that I am missing? Can you please clarify the steps to reproduce. Thank you :bowing_man:

**Comment by mangecoeur** [comment]
@Ben3eeE ok it seems to crop up under very specific circumstances. I've created a test case which for me can trigger the problem:

[test.zip](https://github.com/atom/atom/files/1260242/test.zip)

Steps:

1. unzip the file and open test.md in atom
2. trigger `preview-inline` on BOTH image links
3. Scroll up and down
-> Error seems to be triggered when the top of the block decoration of the first decoration enters or exits the viewable window.
4. Close the first preview
-> Error seems to be triggered when the top of the block decoration of the remainding decoration enters or exits the viewable window. NOTE that if you close the second preview, but keep the first, the error is not triggered `¯\_(ツ)_/¯`
5. Close both preview
-> The error no longer appears.


Hope you can reproduce with these steps. I can't imagine what could be causing such as specific error - maybe something to do with how block decorations are created?

**Comment by Ben3eeE** [comment]
Thank you so much @mangecoeur I can reproduce it now.

It seems to only happen to me when the editor window is narrow.

/cc: @as-cii @nathansobo

**Comment by Ben3eeE** [comment]
I have updated the steps to reproduce in the body of the issue to include the steps you posted @mangecoeur Thank you :bowing_man:

**Comment by as-cii** [comment]
@Ben3eeE @mangecoeur: really amazing job tracking this down. ⚡️ I am working on a fix right now and will close this once it gets merged. Thank you so much!

**Comment by mangecoeur** [comment]
@as-cii glad to help !

**Comment by nathansobo** [comment]
Could this also be connected to #14834? Or would we always add an extra line to measure when a decoration is out of the rendered range?

**Comment by as-cii** [comment]
> Or would we always add an extra line to measure when a decoration is out of the rendered range?

Yeah, I think we would still add an extra line so I am not sure it could trigger that error. I guess we'll be able to confirm this hypothesis by shipping the fix and seeing if we get other reports of #14834.

**Comment by Ben3eeE** [comment]
@as-cii I am still able to reproduce this. With the same steps. I also see an exception from the linter.

[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.22.0-dev-4ce9e1ca1 x64
**Electron**: 1.6.9
**OS**: Microsoft Windows 10 Pro
**Thrown From**: Atom Core


### Stack Trace

Uncaught TypeError: Cannot read property 'id' of undefined

```
At I:\atom_master\atom\out\Atom x64\resources\app\src\text-editor-component.js:1182

TypeError: Cannot read property 'id' of undefined
    at TextEditorComponent.addBlockDecorationToRender (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:1182:67)
    at TextEditorComponent.addDecorationToRender (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:1023:22)
    at decorationsByMarker.forEach (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:992:20)
    at Map.forEach (native)
    at TextEditorComponent.queryDecorationsToRender (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:987:31)
    at TextEditorComponent.updateSyncBeforeMeasuringContent (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:380:16)
    at TextEditorComponent.measureBlockDecorations (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:307:20)
    at TextEditorComponent.updateSync (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:273:16)
    at etch.getScheduler.updateDocument (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:225:46)
    at ViewRegistry.module.exports.ViewRegistry.performDocumentUpdate (I:/atom_master/atom/out/Atom x64/resources/app/src/view-registry.js:187:15)
    at I:/atom_master/atom/out/Atom x64/resources/app/src/view-registry.js:3:65
```

### Commands

```
  2x -4:38.4.0 intentions:highlight (input.hidden-input)
     -4:36.9.0 preview-inline:show (input.hidden-input)
     -4:05.9.0 intentions:highlight (input.hidden-input)
     -4:05.5.0 preview-inline:show (input.hidden-input)
```

### Non-Core Packages

```
atom-beautify 0.30.5 
atom-ide-ui 0.1.11 
atom-material-syntax 1.0.6 
atom-material-ui undefined 
autocomplete-emojis 2.5.0 
busy-signal 1.4.3 
chestnut-light-atom-syntax 0.2.1 
color-picker 2.2.5 
editor-stats 0.17.0 
fizzy 0.21.0 
hey-pane 1.0.0 
highlight-selected 0.13.1 
honor-altgraph 0.1.0 
intentions 1.1.5 
language-latex 1.1.1 
language-markdown 0.25.1 
latex 0.46.0 
less-than-slash 0.17.0 
linter 2.2.0 
linter-js-standard 3.9.3 
linter-ui-default 1.6.7 
markdown-pdf 1.5.4 
minimap 4.29.6 
minimap-autohider 1.5.3 
minimap-find-and-replace 4.5.2 
minimap-highlight-selected 4.6.1 
nebula-syntax 0.4.4 
nebula-ui 0.6.0 
no-caffeine-syntax 0.18.1 
pastel-jelly-syntax 0.2.0 
php-cs-fixer 4.1.0 
pigments 0.40.2 
platformio-ide-terminal 2.5.5 
preview-inline 1.4.7 
project-manager 3.3.5 
project-plus 1.0.0 
remote-edit 1.9.0 
seti-syntax 1.1.3 
seti-ui 1.9.0 
Sublime-Style-Column-Selection 1.7.4 
tab-title 0.3.5 
toggler 0.3.0 
zentabs 0.8.8 
```

[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.22.0-dev-4ce9e1ca1 x64
**Electron**: 1.6.9
**OS**: Microsoft Windows 10 Pro
**Thrown From**: [linter-ui-default](https://github.com/steelbrain/linter-ui-default) package 1.6.7


### Stack Trace

Uncaught TypeError: Cannot match against 'undefined' or 'null'.

```
At I:\atom_master\atom\out\Atom x64\resources\app\src\text-editor-component.js:2370

TypeError: Cannot match against 'undefined' or 'null'.
    at TextEditorComponent.screenPositionForPixelPosition (I:/atom_master/atom/out/Atom x64/resources/app/src/text-editor-component.js:2370:65)
    at getBufferPositionFromMouseEvent (/packages/linter-ui-default/lib/editor/helpers.js:9:50)
    at /packages/linter-ui-default/lib/editor/index.js:168:29)
    at later (/packages/linter/node_modules/sb-debounce/index.js:9:14)
```

### Commands

```
```

### Non-Core Packages

```
atom-beautify 0.30.5 
atom-ide-ui 0.1.11 
atom-material-syntax 1.0.6 
atom-material-ui undefined 
autocomplete-emojis 2.5.0 
busy-signal 1.4.3 
chestnut-light-atom-syntax 0.2.1 
color-picker 2.2.5 
editor-stats 0.17.0 
fizzy 0.21.0 
hey-pane 1.0.0 
highlight-selected 0.13.1 
honor-altgraph 0.1.0 
intentions 1.1.5 
language-latex 1.1.1 
language-markdown 0.25.1 
latex 0.46.0 
less-than-slash 0.17.0 
linter 2.2.0 
linter-js-standard 3.9.3 
linter-ui-default 1.6.7 
markdown-pdf 1.5.4 
minimap 4.29.6 
minimap-autohider 1.5.3 
minimap-find-and-replace 4.5.2 
minimap-highlight-selected 4.6.1 
nebula-syntax 0.4.4 
nebula-ui 0.6.0 
no-caffeine-syntax 0.18.1 
pastel-jelly-syntax 0.2.0 
php-cs-fixer 4.1.0 
pigments 0.40.2 
platformio-ide-terminal 2.5.5 
preview-inline 1.4.7 
project-manager 3.3.5 
project-plus 1.0.0 
remote-edit 1.9.0 
seti-syntax 1.1.3 
seti-ui 1.9.0 
Sublime-Style-Column-Selection 1.7.4 
tab-title 0.3.5 
toggler 0.3.0 
zentabs 0.8.8 
```

**Comment by huyiqun** [comment]
This happens to me when hydrogen (the package, running python) is a activated. Before hydrogen is activated, the script itself does not cause the error. After compiling some code using hydrogen (random amount), and then I just moved my insertion cursor to somewhere, this message suddenly came up, and kept popping for many times instantly no matter what I did (either using the arrow key to move the cursor or scrolling using the mouse). 

Using Atom 1.19.4, Hydrogen 1.20.0
